### ISSUE TITLE
Tattoo support

### ISSUE THREAD

**Issue by Ealb** [elmokki/nationgen#983]
This shouldn't be terribly difficult to add, but there are a few moving pieces I'm not sure of. Mostly, how can we only give tattoos to troops & combatant-type casters? The alternative is not to worry about it and give them to everyone, which would be simple albeit inelegant. Otherwise, it's just a matter of making the appropriate filters and/or straight-up entries in themes for the various tattoo types, and ensuring that nations with tattoos always have activation spells (possibly to include a new one that is battlefield-wide).

**Comment by elmokki** [comment]
There are actually things to consider with tattoos. I mean, Marverni has tribal tattoos for different tribes, but you could also just have the same tattoo for whole nation. Both seem like viable alternatives. In general I think this stuff would probably be somehow added via #freetemplatefilter. There just has to be some way to deliver it to suitable units only. The tattoos could be added in a separate code, but the #freetemplatefilter approach would allow, for example, theming unit looks and names too.

The tattoos are really a more complicated version of the old nationwidefilters which I think are still enabled. The stuff that may just decide that everyone is myopic and has -1 precision. This stuff, along with other stuff, not limited but including the "labs are more expensive" or "reanimation" kind of stuff, should really be rolled to a nation trait system. That's really what nation themes were originally supposed to be. 

My vision, which I probably have rambled on about before, is that after races have been chosen, nations would decide on:

1. Magic paths they must have major access to
2. Distribution of races (not too deterministic, but I'd like to really have distinct "race X is the ruling elite" or "race X is the mages only" type setups appear more reliably)
3. Various national traits (called themes for consistency I guess) that should generally affect rest of the generation. This is the stuff nationwidefilters and nationthemes are currently, except this stuff should also include more complex things. 

Tattoos, for example, would probably have separate single tattoo and multi tattoo traits and they'd affect generation in general by, for example, forcing certain mages. It would also only spawn for nations with certain magic access. Something like reanimation would be more complex and if the program should generate heavier reanimation or #nopreach blood nations, this is the system that could theoretically allow making them at least remotely balanced.

Now that I wrote this all down I have a bit clearer vision of what to code, but still, I have no idea when I can be arsed to even start. Not that this is super complex or anything to get done in a basic way. After all, the race distribution stuff can be ignored for now so it'd really be about making mage generation more controlled and then adding suitable commands for the part 3 to be meaningful.

**Comment by Ealb** [comment]
I've thought about this a bit more, and as far as the current nation generation framework goes, there's some kludgy ways these can work. Right now, there's the N1H1 limited AoE spell and a N1H3 battlefield-wide one I added that takes gems and a long time to cast. But should nature be the only tattoo path? I'd prefer to avoid elemental tattoos (well, unless I could define new ones, in which case I'd LOVE to have eley tats), and astral seems too refined... but there are Death tribals like bone tribe (Ur and indy), and Blood also seems like it'd be reasonable to have tattooed warriors. I could add slightly different variants of the activation spells for death and blood, if we were willing to go that way (I Have Ideas about how to do Blood, though I haven't thought out anything particularly interesting for death). In that case, a nation could have nature magepriests. or some TBD combination of death, blood, and holy if they had tattoos, and it would be reasonable to just handle tattoo spells via guaranteed spells - although computationally I'm not thrilled at the idea of using chanceincs to check every troop in every nation for tattoo commands, especially if there's three sets of spells...

As to how to dole them out, that remains a bit messy but a combination of freetemplates and/or #troopcommands, though that would pretty much demand that nations only have one tattoo. Well, it would also be not terribly difficult to tack some conditional logic into some mounts so if you're primitive and ride a bear, you'd have a chance to get bear tattoos even if your nation wasn't terribly tribal and totemic. Embedding something similar on the wolf/bear/boar/etc cloaks could also work to have unusual tattooed elites.

**Comment by elmokki** [comment]
I have some very mild coding motivation now that I am playing Dom4 again. A lot of the stuff this would benefit from would involve major code changes, but let's see.

My major question is, how should the tattoos be handed out in the first place? As in, if you have multiple different tattoos, should you get "foo tribe" and "boo tribe" on the armor variant level or whether units should have more or less random tattoos from a pool for each unit or something entirely different?

**Comment by elmokki** [comment]
Marverni seems to do it roughly like we do unit tiers or whatever the term is used in code. Similar armor roughly shares a totem.

I think blood and death are fine for tattoos. They are animal tattoos, so nature is obvious, but I can see pretty good thematic reasons for other paths too. Like, using blood of animals to enchant troops with their power or calling upon animal spirits or something. What I don't think these need to necessarily be is holy, not all of them at least.

**Comment by elmokki** [comment]
So, to continue the spam, here's how I'd see it happening, roughly anyway:

1. At generation nation gets theme "tattoos, blood, this tattoo pool"
2. Mages are generated based on requirements of that pool. Perhaps at minimum B1H1 and probably increased chance of mage priests of higher tiers too?
3. Troop templates get a random tattoo, so a nation might have (ignoring other restrictions) plate armored troops with boar tattoo, leather armored with bear tattoo and ring mail armored with snake tattoo. 
4. Sacreds and elites probably have a pretty high chance of rolling a tattoo too.

I think that apart from step 2 this is already doable or at least almost doable without too complicated things.

Does this seem fine? Like I rambled 5 months ago, steps 1-2 can be refined to many other thematic things too later.

**Comment by Ealb** [comment]
Yeah, it does. I was going to use Sauromatia as my model of "how tattoos normally look" while treating Marv as an anomaly, and I still think that might be a good thing to keep in mind, but what you're describing is more generalizable. Marv is like EA Macha in that it's a multi-tribal nation, whereas right now NG makes single-tribe nations when it makes tribal nations, but I've been messing around with something closer to multi-tribal with my has-been-a-work-in-progress-forever Occidentals and their Warrior Societies, so it would be good to have - TBH, even Saur is somewhat multi-tribal. 

Ideally, I'd like to be able to make both Marv-style tattooed nations and Sauro-styled ones. As long as we can define multiple tattoo pools, some of which have very few tattoos and some which have many, that should be doable.

I agree about not all activations needing to be XnHn - I was going to make "must be mage-priests" nature tattoos' conceit for what set them apart. Blood I was thinking of having a relatively-easy-to-cast needs-a-slave-or-two spell that hit a larger AoE or even the whole battlefield but had as a secondary effect an easy-to-resist MR check to cause a wound (or a very-easy-to-resist one that caused bleeding). I'm kinda iffy on that, though, because we can't make it only target units with tattoos. *sigh*

If there was a way to use this idea to procedurally impose a bit more widespread structure within a nation, too, that would be excellent. Right now, we can have a single "school" of mages (plus maybe the tertiary bonus mage), and a school can have 2 or 3 "houses" e.g. Scion/Vessel of the Veiled Typhoon/Gusts/Choir. Looking to Mictlan's model (which should be unusual, but would at a minimum be nice to have for Occidentals), it would be nice if we could have nations that had themed Typhoon/Gusts/Choir sacreds/elites that share some thematic elements with the mages.

As a very-tattoo-specific point of reference, these should primarily be aimed at low-prot nations, as all tattoos include Invulnerability as part of what they do (and Boar is *only* Invul), but (as with Sauro) it's still reasonable for more heavily-armored primtives/tribals to still get them.

**Comment by elmokki** [comment]
Sauromatia could be done with a special pool of tattoo templates. While Marverni would be totally random from a pool, Sauromatia would have templates themselves have some chanceincs to make horse tattoos for horse cavalry and so on. 

I didn't really think about making different tribes per se, but tribes in Dominions are heavily tied to what NationGen does with armors and basesprites: Generate one unit and make copies of it with different weapons. Having a pile of tribes would probably involve a bit of extra code to force certain amount of poses with certain tag or force certain amounts of different themeincs to units or something. Poses are easier. We've talked about them in the past too.

Currently the mage generation is a bit too self contained to my tastes. In the long term I'd like the program to roll for primary mage amount and such at the start and that could be used to generate different sacreds for different mage schools and such. It'd help if theming was made a bit more robust somehow compared to current system of generating stuff and then making up names for it. Names could easily be generated during nation generation rather than after it too and if some interconnected history system is wanted, that can be done later (though it's unlikely to ever actually happen)
